### PR TITLE
Test earliest spring boot version

### PR DESCRIPTION
**Pull request by zeitlinger**
Fixes https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/11357

### CODE REVIEW

**Comment by zeitlinger** [comment]
fails here

```
smoke-tests-otel-starter/spring-boot-2 
gradle dependencies -Ddev=true --stacktrace

FAILURE: Build failed with an exception.

* Where:
Build file '/home/gregor/source/opentelemetry-java-instrumentation/bom/build.gradle.kts' line: 1

* What went wrong:
An exception occurred applying plugin request [id: 'otel.bom-conventions']
> Failed to apply plugin 'otel.bom-conventions'.
   > Build completed with 1 failures.
      > 'void org.gradle.api.internal.artifacts.dsl.LazyPublishArtifact.<init>(org.gradle.api.provider.Provider)'

* Try:
> Run with --info or --debug option to get more log output.
> Get more help at https://help.gradle.org.

```

**Comment by laurit** [comment]
https://scans.gradle.com/s/iluslgkondx7s/failure?expanded-stacktrace=WyIwLTEtMiJd#1

File: settings.gradle.kts [hunk 1]
```diff
@@ -28,6 +28,22 @@ dependencyResolutionManagement {
...
+    create("springBoot2") {
+      val springBoot2Version =
+        gradle.startParameter.projectProperties["springBoot2Version"] ?:
+        (if (latestDepTest) "2.7.18" else "2.4.2")
+      plugin("versions", "org.springframework.boot").version(springBoot2Version)
+    }
+    create("springBoot3") {
+      val springBoot3Version =
+        gradle.startParameter.projectProperties["springBoot3Version"] ?:
+        (if (latestDepTest) "3.3.0" else "3.0.0")
```
Comment (trask) [inline settings.gradle.kts:43]: any idea if we can tie the latest dep version (`3.3.0`) to the artifact somehow so that it will get auto-updated by renovate?

Comment (laurit) [reply settings.gradle.kts]: have you tired using `+` as the version?

File: settings.gradle.kts [hunk 2]
```diff
@@ -30,25 +30,19 @@ dependencyResolutionManagement {
...
+        val version =
+          gradle.startParameter.projectProperties["${name}Version"]
+            ?: (if (latestDepTest) "+" else minVersion)
+        plugin("versions", "org.springframework.boot").version(version)
+      }
     }
+    // r2dbc is not compatible with earlier versions
+    addSpringBootCatalog("springBoot2", "2.6.15")
+    addSpringBootCatalog("springBoot3", "3.0.0")
+    addSpringBootCatalog("springBoot32", "3.2.0")
```
Comment (trask) [inline settings.gradle.kts:45]: maybe pass in an extra parameter, e.g. to test against the latest 2.x version?
```suggestion
    addSpringBootCatalog("springBoot2", "2.6.15", "2.+")
    addSpringBootCatalog("springBoot3", "3.0.0", "3.1.+")
    addSpringBootCatalog("springBoot32", "3.2.0", "3.+")
```

**Comment by zeitlinger** [comment]
@jeanbisutti please check

File: .github/workflows/reusable-native-tests.yml [hunk 3]
```diff
@@ -32,5 +32,5 @@ jobs:
           # therefore we're starting a Kafka container manually for the tests
           docker compose -f .github/graal-native-docker-compose.yaml up -d
           # don't wait for startup - gradle compile takes long enough
-          ./gradlew nativeTest
+          ./gradlew -PtestLatestDeps=true nativeTest
```
Comment (laurit) [inline .github/workflows/reusable-native-tests.yml:35]: Since latest dep tests can fail without us doing anything wrong we have made it so that when latest dep test fails it won't prevent the PR form being merged. Perhaps we should make sure that native test failures are also treated that way.

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts [hunk 4]
```diff
@@ -7,7 +7,7 @@ base.archivesName.set("opentelemetry-spring-boot")
 group = "io.opentelemetry.instrumentation"
 
 val versions: Map<String, String> by project
```
Comment (laurit) [inline instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:9]: is this unused now?

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts [hunk 5]
```diff
@@ -7,7 +7,7 @@ base.archivesName.set("opentelemetry-spring-boot")
...
-val springBootVersion = versions["org.springframework.boot"]
+val springBootVersion = "2.7.18"
```
Comment (laurit) [inline instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:10]: I guess this version won't be updated any more?

Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:10]: yes - any idea how to make it work?

File: .github/workflows/reusable-native-tests.yml (hunk 3 above)
Comment (zeitlinger) [reply .github/workflows/reusable-native-tests.yml:35]: native tests are only run when you change something in spring - so maybe it's ok as is. 

alternatively, we could run the native tests twice - but it seems overkill

File: smoke-tests-otel-starter/spring-boot-3.2/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,76 @@
+plugins {
+  id("otel.java-conventions")
+  alias(springBoot32.plugins.versions)
+  id("org.graalvm.buildtools.native")
+}
+
+description = "smoke-tests-otel-starter-spring-boot-3.2"
```
[the remark this answers is not in the document]
Comment (jeanbisutti) [reply smoke-tests-otel-starter/spring-boot-3.2/build.gradle.kts:7]: I would also rename the `spring-boot-3.2` folder into `spring-boot-latest`.

[the remark this answers is not in the document]
Comment (zeitlinger) [reply smoke-tests-otel-starter/spring-boot-3.2/build.gradle.kts:7]: if a now feature is added that works only in 3.3, then we should not mix it with 3.2

so I'd leave it as is

[the remark this answers is not in the document]
Comment (jeanbisutti) [reply smoke-tests-otel-starter/spring-boot-3.2/build.gradle.kts:7]: Ah, I thought it was for testing.

[the remark this answers is not in the document]
Comment (jeanbisutti) [reply smoke-tests-otel-starter/spring-boot-3.2/build.gradle.kts:7]: Oh, I see that there are now the `spring-boot-3` and ` spring-boot-3-2` projects in `smoke-tests-otel-starter`.


> if a now feature is added that works only in 3.3, then we should not mix it with 3.2

It has the advantage of making it explicit in the tests that a feature only works from Spring Boot 3.2. The drawback back is to add a test project only for this.

[the remark this answers is not in the document]
Comment (jeanbisutti) [reply smoke-tests-otel-starter/spring-boot-3.2/build.gradle.kts:7]: More test projects is more difficult to understand and maintain. Someone may inadvertently add a test in the `spring-boot-3-2` folder for a feature that works with all versions of Sprint Boot 3.

[the remark this answers is not in the document]
Comment (zeitlinger) [reply smoke-tests-otel-starter/spring-boot-3.2/build.gradle.kts:7]: I'd suggest we revisit this question once we actually have the case.

Right now, 3.2 is easy to understand - it tests features where the earliest version is 3.2

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts [hunk 7]
```diff
@@ -6,8 +6,7 @@ plugins {
 base.archivesName.set("opentelemetry-spring-boot")
 group = "io.opentelemetry.instrumentation"
 
-val versions: Map<String, String> by project
-val springBootVersion = versions["org.springframework.boot"]
+val springBootVersion = "2.7.18" // AutoConfiguration is added in 2.7.0, but can be used with older versions
```
Comment (zeitlinger) [inline instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:9]: let's see if we find a better solution in a later PR

**Review by jeanbisutti — APPROVED** [review]

File: smoke-tests-otel-starter/spring-smoke-testing/build.gradle.kts [hunk 8]
```diff
@@ -2,7 +2,7 @@ import org.springframework.boot.gradle.tasks.bundling.BootJar
 
 plugins {
   id("otel.java-conventions")
-  id("org.springframework.boot") version "2.7.18"
+  id("org.springframework.boot") version "2.6.15"
```
Comment (laurit) [inline smoke-tests-otel-starter/spring-smoke-testing/build.gradle.kts:5]: why are you changing the version here, won't renovate immediately change it back?

File: smoke-tests/images/spring-boot/build.gradle.kts [hunk 9]
```diff
@@ -5,12 +5,12 @@ plugins {
   id("otel.java-conventions")
 
   id("com.google.cloud.tools.jib")
-  id("org.springframework.boot") version "2.7.18"
+  id("org.springframework.boot") version "2.6.15"
```
Comment (laurit) [inline smoke-tests/images/spring-boot/build.gradle.kts:8]: same here, won't renovate update it automatically?

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 7 above)
Comment (laurit) [inline instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:9]: I don't really understand what this comment means

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts [hunk 10]
```diff
@@ -150,7 +149,7 @@ tasks {
     dependsOn(testing.suites)
   }
 
-  compileTestJava {
+  compileJava {
```
Comment (laurit) [inline instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:152]: why do you need to do this?

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 7 above)
Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:9]: AutoConfiguration is a class that we use to specify the order of autoconfig classes

even though it's added it 2.7.0, the tests run successfully with 2.6

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts [hunk 11]
```diff
@@ -22,7 +21,7 @@ sourceSets {
 }
 
 dependencies {
-  implementation("org.springframework.boot:spring-boot-autoconfigure:$springBootVersion")
+  compileOnly("org.springframework.boot:spring-boot-autoconfigure:$springBootVersion")
```
Comment (laurit) [inline instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:24]: why is this needed?

File: instrumentation/spring/starters/spring-boot-starter/build.gradle.kts [hunk 12]
```diff
@@ -5,12 +5,11 @@ plugins {
...
 
-val versions: Map<String, String> by project
-val springBootVersion = versions["org.springframework.boot"]
+val springBootVersion = "2.6.15"
 
 dependencies {
-  api("org.springframework.boot:spring-boot-starter:$springBootVersion")
-  api("org.springframework.boot:spring-boot-starter-aop:$springBootVersion")
+  compileOnly("org.springframework.boot:spring-boot-starter:$springBootVersion")
+  compileOnly("org.springframework.boot:spring-boot-starter-aop:$springBootVersion")
```
Comment (laurit) [inline instrumentation/spring/starters/spring-boot-starter/build.gradle.kts:12]: why is this needed?

File: smoke-tests-otel-starter/spring-smoke-testing/build.gradle.kts (hunk 8 above)
Comment (zeitlinger) [reply smoke-tests-otel-starter/spring-smoke-testing/build.gradle.kts:5]: will add to renovate rules

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 10 above)
Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:152]: I got a warning without it

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 11 above)
Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:24]: the spring version was used in the tests - but the user should include the spring starter directly

File: instrumentation/spring/starters/spring-boot-starter/build.gradle.kts (hunk 12 above)
Comment (zeitlinger) [reply instrumentation/spring/starters/spring-boot-starter/build.gradle.kts:12]: the spring version was used in the tests - but the user should include the spring starter directly

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 11 above)
Comment (trask) [inline instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:24]: would it work to change the base version to the real base version (2.6.0?), and change this to

```suggestion
  // needed to compile against optional AutoConfiguration annotation
  compileOnly("org.springframework.boot:spring-boot-autoconfigure:2.7.0")
```

and have the primary test be against the base version

and if you want to test the AutoConfiguration annotation code, then add a test suite to test against 2.7.0

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 10 above)
Comment (trask) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:152]: an existing warning or a new warning related to this PR?

File: .github/renovate.json5 [hunk 13]
```diff
@@ -116,7 +116,7 @@
         "org.slf4j:slf4j-api",
         "org.springframework.boot",
         "org.springframework.boot:spring-boot-dependencies"],
-      "matchUpdateTypes": ["major"],
+      "matchUpdateTypes": ["major", "minor"],
```
Comment (trask) [inline .github/renovate.json5:119]: i'm not sure we want this, it will prevent renovate from updating us from Spring Boot 3.3 to 3.4 in the future

File: .github/workflows/reusable-native-tests.yml (hunk 3 above)
Comment (trask) [reply .github/workflows/reusable-native-tests.yml:35]: > we could run the native tests twice

I think this makes sense for the daily native CI, it gives us coverage and confidence that we haven't broken the old, and that we work with the new

for PR build I'd say it would be ideal to only run against the old version for reason @laurit gives above

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 11 above)
Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:24]: - the smoke test is run against 2.6
- 2.6.0 doesn't work, it contains some issues

> and if you want to test the AutoConfiguration annotation code, then add a test suite to test against 2.7.0

it's tested with latest version

File: .github/renovate.json5 (hunk 13 above)
Comment (zeitlinger) [reply .github/renovate.json5:119]: no, spring 3 uses https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/dc4330e0a3060bd7d8c4090ad0b8fc4727e68113/settings.gradle.kts#L43-L45

I added a comment explaining this

File: settings.gradle.kts [hunk 14]
```diff
@@ -41,7 +41,8 @@ dependencyResolutionManagement {
     }
     // r2dbc is not compatible with earlier versions
     addSpringBootCatalog("springBoot2", "2.6.15", "2.+")
-    addSpringBootCatalog("springBoot3", "3.0.0", "3.+")
+    // spring boot 3.0 is not compatible with graalvm native image
+    addSpringBootCatalog("springBoot3", "3.1.0", "3.+")
```
Comment (trask) [inline settings.gradle.kts:45]: does this make more sense?
```suggestion
    addSpringBootCatalog("springBoot31", "3.1.0", "3.1.+")
```

File: .github/workflows/reusable-native-tests.yml [hunk 15]
```diff
@@ -32,5 +32,9 @@ jobs:
           # therefore we're starting a Kafka container manually for the tests
           docker compose -f .github/graal-native-docker-compose.yaml up -d
           # don't wait for startup - gradle compile takes long enough
-          ./gradlew nativeTest
+          PROFILES=""
+          if [ -n "$LATEST_DEPS" ]; then
+              PROFILES="-PtestLatestDeps=true"
+          fi
+          ./gradlew $PROFILES nativeTest
```
Comment (trask) [inline .github/workflows/reusable-native-tests.yml:39]: I'd suggest instead adding explicit "workflow input" at the top, e.g.

```
on:
  workflow_call:
    inputs:
      skip-native-tests:
        type: boolean
        required: false
      test-latest-deps:
        type: boolean
        required: false
```

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 11 above)
Comment (trask) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:24]: sorry, I didn't follow how it works and doesn't work at the same time? thanks

> * the smoke test is run against 2.6
> * 2.6.0 doesn't work, it contains some issues

File: settings.gradle.kts (hunk 14 above)
Comment (zeitlinger) [reply settings.gradle.kts:45]: the latest version should still be 3.3

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 11 above)
Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:24]: seems it needs 2.7 to compile, but not at runtime

Comment (trask) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:24]: can we have compileOnly on 2.7, and testImplementation (or testLibrary) on 2.6?

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 10 above)
Comment (trask) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:152]: what's the warning?

File: settings.gradle.kts (hunk 14 above)
Comment (trask) [reply settings.gradle.kts]: isn't testing of the latest version covered below by

```
    addSpringBootCatalog("springBoot32", "3.2.0", "3.+")
```

Comment (zeitlinger) [reply settings.gradle.kts]: no, `springBoot32` is used by the module that has 3.2 as min version

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 11 above)
Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts]: tried that - results is runtime errors (` java.lang.NoSuchMethodError: 'reactor.core.publisher.Mono reactor.core.publisher.Mono.subscriberContext(reactor.util.context.Context)'`) which often happen if you have different spring versions

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 10 above)
Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts]: apprently it wasn't needed :wink:

File: instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts (hunk 11 above)
Comment (trask) [reply instrumentation/spring/spring-boot-autoconfigure/build.gradle.kts:24]: thanks for trying ❤️

**Review by trask — APPROVED** [review]
